src/Xamarin.Android.Build.Tasks/Utilities/TypeMapGenerator.cs

```diff
@@ -315,104 +466,50 @@ void OutputModule (byte[] moduleUUID, ModuleData moduleData)
-		//  [Java type name width]            # unsigned 32-bit integer, 4 bytes
-		//  [Assembly name size]              # unsigned 32-bit integer, 4 bytes
+		//  [Magic string]                    # XATS
+		//  [Format version]                  # 32-bit unsigned integer, 4 bytes
+		//  [Entry count]                     # 32-bit unsigned integer, 4 bytes
+		//  [Java type name width]            # 32-bit unsigned integer, 4 bytes
+		//  [Managed type name width]         # 32-bit unsigned integer, 4 bytes
+		//  [Assembly name size]              # 32-bit unsigned integer, 4 bytes
 		//  [Assembly name]                   # Non-null terminated assembly name
 		//  [Java-to-managed map]             # Format described below, [Entry count] entries
 		//  [Managed-to-java map]             # Format described below, [Entry count] entries
-		//  [Managed-to-java duplicates map]  # Map of unique managed IDs which point to the same Java type name (might be empty)
 		//
 		// Java-to-managed map format:
 		//
-		//    [Java type name]<NUL>[Managed type token ID]
+		//    [Java type name]<NUL>[Managed type table index]
 		//
 		//  Each name is padded with <NUL> to the width specified in the [Java type name width] field above.
 		//  Names are written without the size prefix, instead they are always terminated with a nul character
 		//  to make it easier and faster to handle by the native runtime.
 		//
-		//  Each token ID is an unsigned 32-bit integer, 4 bytes
-		//
-		//
-		// Managed-to-java map format:
-		//
-		//    [Managed type token ID][Java type name table index]
-		//
-		//  Both fields are unsigned 32-bit integers, to a total of 8 bytes per entry. Index points into the
-		//  [Java-to-managed map] table above.
+		//  Each [Managed type table index] is an unsigned 32-bit integer, 4 bytes
 		//
-		// Managed-to-java duplicates map format:
 		//
-		//  Format is identical to [Managed-to-java] above.
+		// Managed-to-java map is identical to the [Java-to-managed] table above, with the exception of the index
+		// pointing to the Java name table.
 		//
-		void OutputModule (BinaryWriter bw, byte[] moduleUUID, ModuleData moduleData)
+		void OutputModule (BinaryWriter bw, string moduleName, ModuleDebugData moduleData)
 		{
 			bw.Write (moduleMagicString);
 			bw.Write (TypeMapFormatVersion);
-			bw.Write (moduleUUID);
-
-			var javaNames = new Dictionary<string, uint> (StringComparer.Ordinal);
-			var managedTypes = new Dictionary<uint, uint> ();
-			int maxJavaNameLength = 0;
-
-			foreach (TypeMapEntry entry in moduleData.Types) {
-				javaNames.Add (entry.JavaName, entry.Token);
-				if (entry.JavaNameLength > maxJavaNameLength)
-					maxJavaNameLength = entry.JavaNameLength;
-
-				managedTypes.Add (entry.Token, 0);
+			bw.Write (moduleData.JavaToManagedMap.Count);
+			bw.Write (moduleData.JavaNameWidth);
+			bw.Write (moduleData.ManagedNameWidth);
+			bw.Write (moduleName.Length);
+			bw.Write (outputEncoding.GetBytes (moduleName));
```

Comment:
Rando fear, but nothing requires that assembly names be ASCII, so if `moduleName` contains non-ASCII characters, `outputEncoding.GetBytes(moduleName).Length` will not equal `moduleName.Length`.

This should probably instead be:

```csharp
byte[] moduleNameBytes = outputEncoding.GetBytes (moduleName);
bw.Write (moduleNameBytes.Length);
bw.Write (moduleNameBytes);
```

Reply:
Good call, want me to update it here or in a new PR?